lib/types.ts

```diff
@@ -0,0 +1,85 @@
+import type * as typesObject from './constant/types'
+import type {ValueOf} from 'type-fest'
+
+export interface Point {
+  line: number
+  column: number
+  offset: number
+  _index: number
+  _bufferIndex: number
+}
+export interface Token {
+  start: Point
+  end: Point
+  type: unknown
+  contentType: string
+  previous: Token
+  next: Token
+  _size: number
+  _break?: boolean
+  _tokenizer?: unknown
+  _events?: Event[]
+  _subevents?: Event[]
+  _contentTokenized?: unknown
+  _close?: unknown
+  _marker?: unknown
+  _open?: unknown
+  _side?: number
+  _tabs?: unknown
+}
+
+export type Type = ValueOf<typeof typesObject>
+
+export type Event = [name: string, token: Token, context: unknown]
+
+export interface Effects {
+  enter: (type: Type) => Token
+  consume: (code: number) => Token
+  exit: (type: Type) => Token
+  attempt: (
+    one: unknown,
+    two: unknown,
+    three?: unknown
+  ) => (code: number) => unknown
+  interrupt: (
+    one: unknown,
+    two: unknown,
+    three: unknown
+  ) => (code: number) => unknown
+  check: (
+    one: unknown,
+    two: unknown,
+    three?: unknown
+  ) => (code: number) => unknown
+  lazy: (one: unknown, ok: Okay, nok: NotOkay) => void
+}
+
+export interface Okay {
+  (code: number): void
+}
+
+export interface NotOkay {
+  (code: number): void
+}
+
+export interface Parser {
+  hooks: {
+    [key: string]: unknown
+  }
+  flow: (something: unknown) => unknown
+  defined: unknown[]
+}
+
+export interface TokenizerThis {
+  events: Event[]
+  interrupt: unknown
+  lazy: unknown
+  containerState: {
+    marker: number
+    type: Type
+    initialBlankLine: unknown
+    size: number
+    _closeFlow: unknown
+    furtherBlankLines: unknown
```

Comment:
containerState is used in the document (container) tokenizer, because generally we can parse in one go (e.g., an atx heading), but for lists and block quotes we parse a part of it (e.g., `> `), then we do the rest of the line, and then at the next line look for another marker. As we parse separate “runs” of content, information needs to be stored somewhere, and I came up with this.

All properties are used by lists

_closeFlow is a way to communicate from lists to the tokenizer that the flow is closed, but we do continue the list. That’s useful when finding a new list item, because the previous one needs to be closed, but the list remains open.